pygradle-plugin/src/main/groovy/com/linkedin/gradle/python/plugin/AbstractPluginBase.groovy

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2016 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.gradle.python.plugin
+
+import com.linkedin.gradle.python.PythonExtension
+import com.linkedin.gradle.python.util.values.PyGradleConfiguration
+import com.linkedin.gradle.python.util.values.PyGradleTask
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.artifacts.Configuration
+
+import static com.linkedin.gradle.python.util.values.PyGradleConfiguration.*
+
+abstract class AbstractPluginBase implements Plugin<Project> {
+
+    public static final String PYTHON_EXTENSION_NAME = 'python'
+
+    public Project project
+    public PythonExtension settings
+
+    @Override
+    void apply(Project target) {
+        this.project = target
+
+        project.plugins.apply('base')
+
+        settings = addGetExtensionLocal(PYTHON_EXTENSION_NAME, PythonExtension)
+        applyTo(target)
+    }
+
+    abstract void applyTo(Project project)
+
+    /**
+     * This adds the plugin to the local project only if it doesn't already exist
+     * @param project project object
+     * @param plugin plug class to add
+     * @return
+     */
+    protected <T extends Plugin> T addPluginLocal(Class<T> plugin) {
+        if (!project.plugins.withType(plugin)) {
```

Comment:
because in my suite, it used to fail all the time with "plugin already exists".  Tracking these down as we stacked plugin on top of plugin finally got us to give up and we wrote the abstraction.  solved the problem and cleaned up the DSL a ton for readability.